[App] Use the existing files when saving an existing project.

SebDieBln opened the pull request:
When saving an existing project the old behaviour was
- delete the backup file (if existent)
- rename the project file to be the backup file
- recreate the project file and write the project data to it

This lead to the loss of special permissions of the project file (it got the default permissions) and of the backup file (it got the permissions of the former project file). Additionally (at least on Windows) the icons for the files kept reappearing in their default locations in the explorer.

I did not test it, but I assume failures when the backup file was located in a folder where the user was not allowed to delete or create files (e.g. permissions `rwxr-xr-x` and not being the owner) or when the project/backup files were symlinks.

This PR fixes this by
- copying the contents of the project file to the existing backup file
- writing the project data to the existing project file
- creating new files only when needed

SebDieBln committed:
use existing files when saving existing project

m-kuhn commented:
I was thinking about an alternative approach like this:

 * Create empty backup file (or copy if it helps for permissions)
 * Write to backup file
 * After success rename and overwrite master file (or swap the two files to still keep a backup from before)

At the moment, if QGIS crashes during writing a project file, the file will be empty and a user has to know that a backup file exists and that it is safe to rename this file and use it.

SebDieBln commented:
I think the key is minimizing the likelihood of
> if QGIS crashes during writing a project file, the file will be empty

and to give the user a hint
> that a backup file exists and that it is safe to rename this file and use it

I would like to adress these points by:
* Generating all the XML data first, then performing the backup and write the new data to the project file. In case of a crash the old project will be untouched. The time between truncating the project file and writing the new data is kept as small as possible.
* Make QGIS ask the user whether to load the backup file when a project file is corrupted.

Would that be ok for you?

m-kuhn commented:
Crashes during saving the project file were the reason for the backup IIRC and a plugin may crash which we cannot prevent..

> The time between truncating the project file and writing the new data is kept as small as possible.

File renaming is an atomic operation (on decent file systems), I wouldn't even truncate.

> Make QGIS ask the user whether to load the backup file when a project file is corrupted.

With the approach of writing to separate file and then renaming, the only cause for this scenario to happen is probably a user messing manually with the .qgs file. But it's nice to be prepared for that :)

SebDieBln committed:
change project/backup file only after preparing the data
This avoids corrupted project files when crashing during the preparation of the XML data to write to the project file, e.g. due to a faulty plugin.

SebDieBln committed:
prompt the user to load backup when load project fails
If an error occurs while loading the project file, prompt the user to load the backup file instead, if it exists.

SebDieBln commented:
Ok, here are my additions:
* The project and backup files are only touched __after__ the complete XML data is prepared. So if QGIS crashes during this process the project file is not corrupted.
* When trying to load a corrupted project file and a backup file exists, the user is prompted to load the backup.

> File renaming is an atomic operation (on decent file systems), I wouldn't even truncate.

The whole point of this PR is to avoid renaming files when actually only the contents change. And while I agree that
```c++
imp_->file.open( QIODevice::WriteOnly | QIODevice::Truncate );
QTextStream projectFileStream( &imp_->file );
doc->save( projectFileStream, 2 );
imp_->file.close();
```
is not atomic, the room for errors that leave us with a corrupted project file is highly reduced compared to the current situation in _master_.

m-kuhn commented:
I agree that what you explain sounds better than the current situation, but the atomic one still appears to be failure safe in comparison to almost failure safe ;)

Can you explain again why

> The whole point of this PR is to avoid renaming files when actually only the contents change.

m-kuhn commented:
Would copying (instead of creating from scratch), replace contents in other file and renaming back to original avoid losing special permissions?

SebDieBln commented:
Renaming has side effects compared to changing the contents. Here are the ones that come to my mind, __but there might be more__:
* The icon moves to the default position within Windows Explorer, e.g. on the desktop. (This is actually the reason for creating the PR in the first place.)
* The permissions are changed to the default. This might be partially overcome by copying instead of creating from scratch, but is closely related to the next point.
* The ownership is changed to the default. Try opening and saving a project that is owned by another user.
* It won't work in folders where the user does not have the right to add/delete files. Try opening and saving a project (with a backup file) in a folder that is `chmod 555`. The backup file will not be updated.
* Symlinks will be lost. Try opening and saving a symlink to a project instead of the project itself. You should end up with a regular file instead of modifying the file pointed to by the symlink.
* I have no idea what it does to contexts, e.g. security contexts like SElinux.

Note that I do not use QGIS under Linux, so most of these points are based on my experience with Linux and some tests on the command line, however ACLs under Windows do not react well to renaming either.

m-kuhn commented on src/core/qgsproject.cpp:
```diff
@@ -779,11 +779,11 @@ bool QgsProject::addLayer( const QDomElement &layerElem, QList<QDomNode> &broken
    Please note that most of the contents were copied from qgsproject
 */
-bool QgsProject::read( QFileInfo const &file )
+bool QgsProject::read( QFileInfo const &file, int * errorCode )
```
Could this be an enum instead?

SebDieBln replied:
Yes.

m-kuhn commented:
Sounds like troubles for non-default situations indeed. So let's go for your proposal instead.

SebDieBln commented on src/core/qgsproject.h:
```diff
@@ -72,6 +72,14 @@ class CORE_EXPORT QgsProject : public QObject
+    {
+      NoError,
+      IOError,       // data could not be read/written
+      ParseError,    // data could not be parsed
```
I am open for suggestions regarding the naming of the constants.

m-kuhn replied:
The code names look good to me, but the enum should be CamelCase to conform with the rest.

SebDieBln replied:
CamelCase it is now.

m-kuhn commented:
Looks good to me,

Can you rebase against master?

Thanks a lot!!

SebDieBln commented:
> Can you rebase against master?

Done.

m-kuhn commented:
Thanks
